### ISSUE TITLE
`drush cr` doesn't work if Drupal is symlinked in by Composer (Drush 12)

### ISSUE THREAD

**Issue by joachim-n** [drush-ops/drush#4712]
**Describe the bug**

If Drupal core is symlinked in by Composer (see https://github.com/joachim-n/drupal-core-composer) then `drush cr` fails with an error message about no hash salt.

This is because DrupalKernel and Settings get initialised with the wrong app root.

This is because CacheCommands::rebuild() is using the DRUPAL_ROOT constant that gets set by includes/bootstrap.inc, and it shouldn't, because that constant is assuming the Drupal package is in the standard place. (See https://www.drupal.org/project/drupal/issues/1792310 for more on this.)

The fix is actually very simple. By some magic I don't understand, the PHP current directory is already set to the app root at this point. So we can simply use getcwd().

MR coming.

**To Reproduce**

Installed Drupal with https://github.com/joachim-n/drupal-core-composer. Did `drush cr`.

**Expected behavior**

It works.

**Actual behavior**

Error message.

**Workaround**

Clear it in the UI, probably.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 10.x
| Drupal version? | 9.2
| PHP version     | 7.4
| OS?             | Mac

**Additional information**

**Comment by joachim-n** [comment]
Of course it THEN does a cache rebuild with the wrong app root, because of https://git.drupalcode.org/project/drupal/-/merge_requests/182#note_17888. But that's a problem in Drupal core, not in Drush.

**Comment by weitzman** [comment]
Fix constant in Drupal please

**Comment by joachim-n** [comment]
Could this be reopened please? PR #5427 has broken `joachim-n/drupal-core-development-project`.

The fix in Drupal core is not going to land any time soon.

**Comment by weitzman** [comment]
Maybe you could make an MR which shows what you want to change? Note that Drush12 finds Drupal root via Composer Runtime API now, and not using DrupalFinder. See https://github.com/drush-ops/drush/blob/4e71e538d6383092ea03c79d47f20c8a596ba5a7/src/DrupalFinder/DrushDrupalFinder.php#L26-L27
https://github.com/weitzman/drush/blob/4e71e538d6383092ea03c79d47f20c8a596ba5a7/src/DrupalFinder/DrushDrupalFinder.php#L26-L27

**Comment by weitzman** [comment]
Note that this change mens that we can successfully find drupal/core when it is a Path repository located in an unusual spot.

**Comment by joachim-n** [comment]
My laptop isn't ready to run Drupal 10 or Drush 12, unfortunately - I need to find time to upgrade my PHP first. So I'm fumbling in the dark somewhat here.

> Note that Drush12 finds Drupal root via Composer Runtime API now, and not using DrupalFinder.

👍 

That's great! 

The correct app root also needs to be passed in to DrupalKernel, otherwise it will look in the wrong place for contrib modules.

See https://github.com/joachim-n/drupal-core-development-project/blob/main/drush/Commands/core_development/DevelopmentProjectCommands.php -- all the changes I needed to make to the Drush cache clear command are marked in comments.

**Comment by greg-1-anderson** [comment]
> Replaces the cache:rebuild command to correctly work with symlinked Drupal.

Extending core Drush command classes is no longer allowed in Drush 12.

**Comment by greg-1-anderson** [comment]
`DevelopmentProjectCommands` does not appear to need to extend `CacheCommands`, or at least, if there is a reason that it does, I failed to notice it. Maybe you could try [extending / implementing the same things CacheCommands does](https://github.com/drush-ops/drush/blob/12.x/src/Commands/core/CacheCommands.php#L30)?

I didn't compare your version of `cr` with the one in Drush core, so I don't know if this is something that we could just adopt so that you do not need to replace the `cr` command.

**Comment by joachim-n** [comment]
> Extending core Drush command classes is no longer allowed in Drush 12.

Yes, that's why the Composer template is broken with Drush 12 - see https://github.com/joachim-n/drupal-core-development-project/issues/19.

> I didn't compare your version of cr with the one in Drush core, so I don't know if this is something that we could just adopt so that you do not need to replace the cr command.

If you're now getting the Drupal root from the Composer API, then I think the only further change that Drush would need to make is to pass the app root in to DrupalKernel, like my subclass does:

```
    $kernel = new DrupalKernel('prod', $class_loader, TRUE, $app_root);
```

That's needed because DrupalKernel guesses the app root incorrectly.

**Comment by greg-1-anderson** [comment]
Seems like this is a safe change to make in Drush. @weitzman, what do you think?

**Comment by weitzman** [comment]
AFAICT, [we already do that](https://github.com/drush-ops/drush/blob/8664a4f9d380f62df99b142b19475a65a042c57e/src/Boot/DrupalBoot8.php#L219). You added that in https://github.com/drush-ops/drush/pull/4585 :)

**Comment by joachim-n** [comment]
Aha!

Ok, in which case it *might* be enough to bump the Drush version in drupal-core-development-project and remove the class that overrides the cache clear command. I'll experiment and report back.

**Comment by greg-1-anderson** [comment]
🤞

**Comment by bhanu951** [comment]
> it might be enough to bump the Drush version in drupal-core-development-project and remove the class that overrides the cache clear command. I'll experiment and report back.

It doesn't work. It produces following error on site install.
https://www.drupal.org/project/drupal/issues/3372954

**Comment by weitzman** [comment]
This issue is about Drush 12. That issue is using Drupal 9 which wont work with Drush 12.

**Comment by bhanu951** [comment]
That issue is not using Drupal 9, its using Drupal 11.x with Drush 12.

Am I missing something?

```
drush st
Drupal version   : 11.0-dev
Site URI         : https://drupal-contrib.ddev.site
DB driver        : mysql
DB hostname      : db
DB port          : 3306
DB username      : db
DB name          : db
Database         : Connected
Drupal bootstrap : Successful
Default theme    : olivero
Admin theme      : claro
PHP binary       : /usr/bin/php8.1
PHP config       : /etc/php/8.1/cli/php.ini
PHP OS           : Linux
PHP version      : 8.1.16
Drush script     : /var/www/html/vendor/bin/drush
Drush version    : 12.1.1.0
Drush temp       : /tmp
Drush configs    : /var/www/html/vendor/drush/drush/drush.yml
Install profile  : standard
Drupal root      : /var/www/html/web
Site path        : sites/default
Files, Public    : sites/default/files
Files, Temp      : /tmp

```

**Comment by bhanu951** [comment]
Here is a repo to reproduce the issue. With instructions to setup.
https://github.com/bhanu951/drupal-core-development-project/tree/master#core-development-using-ddev
Make sure to setup with 10.x before switching to 11.x.

**Comment by joachim-n** [comment]
There's no need to switch to 11.x - that's muddying the waters, as it looks like Drupal 11 doesn't install with this template (https://www.drupal.org/project/drupal/issues/3372954).

I can reproduce this with Drupal 10 as follows:

1. composer create-project joachim-n/drupal-core-development-project
2. cd repos/drupal
3. git co 10.1.x
4. cd -
5. composer update
6. drush si
7. composer require drupal/admin_toolbar
8. drush en -y drupal/admin_toolbar
9. composer require drush/drush ^12 -W
10. rm drush/Commands/core_development/DevelopmentProjectCommands.php
11. drush cc drush # needed because otherwise Drush looks for the class that's just been deleted
12. drush cr
13. visit web/admin/modules - admin toolbar is no longer showing. 


EDIT: I pasted the wrong project template name! It's supposed to be joachim-n/drupal-core-development-project.

**Comment by weitzman** [comment]
In that list I dont see how you got DevelopmentProjectCommands.php file.

Further, what do we mean by "is symlinked by composer" in the Title of this issue. is drupal/core a path repository? If not I dont see how composer core is making the symlink. Maybe its your composer plugin making a symlink which is unknown to the composer autoloader. If so, then the Composer Runtime API will be of no help.

**Comment by joachim-n** [comment]
Sorry, I pasted the wrong Composer template name when I posted the steps -- it's joachim-n/drupal-core-development-project. Which is how drupal/core is a path repo, with Composer making the symlink when it installs the package, and also how DevelopmentProjectCommands is present.

**Comment by weitzman** [comment]
I stopped after Step5, after seeing the messages below. I dont think you expected a downgrade there?
```
- Downgrading drupal/core-vendor-hardening (11.x-dev => 10.1.x-dev): Source already present
  - Downgrading drupal/core-project-message (11.x-dev => 10.1.x-dev): Source already present
  - Downgrading drupal/drupal (11.x-dev => 10.1.x-dev): Source already present
  - Downgrading drupal/core (11.x-dev => 10.1.x-dev): Source already present
```

Reviewing the codebase, your /repos/drupal/core is a Path repository so Composer knows about it fine. Your /web/core directory symlink is unknown the Composer Runtime API so Drush is going to find the Drupal root /repos/drupal. Perhaps you could coax it to be found at /web via the --root option.

**Comment by weitzman** [comment]
> CacheCommands::rebuild() is using the DRUPAL_ROOT constant 

Thats not true in Drush 12, as far as I can tell.  This issue has twists and turns and not much useful information. I suggest closing it and making a new issue once the Drush change request becomes more clear.

**Comment by bhanu951** [comment]
> Here is a repo to reproduce the issue. With instructions to setup. https://github.com/bhanu951/drupal-core-development-project/tree/master#core-development-using-ddev Make sure to setup with 10.x before switching to 11.x.

@weitzman if you are having issues with @joachim-n instructions maybe you can try with this instructions. 
Its a simple ddev configuration.

As @joachim-n mentioned you can test Drush 12 on Drupal core 10.x no need to upgrade to 11.x as well.

**Comment by joachim-n** [comment]
> I dont think you expected a downgrade there?

That's normal. The clone at repos/drupal gets the default branch, which at the moment will be 11.x. So when Composer installs the project you get all the 11.x packages. We want to test this issue on 10.x, because 11 has additional problems.

**Comment by joachim-n** [comment]
Ok I've found the problem. In https://github.com/joachim-n/drupal-core-development-project/blob/fd0eba4c931266003a446a6317ef5c42f030865a/drush/Commands/core_development/DevelopmentProjectCommands.php#L64 we have a replacement for Drupal core's `drupal_rebuild()`, because **that** instantiates `DrupalKernel` without passing in the app root. `DrupalKernel` then guesses the app root incorrectly.

So, the drush change request is one of:

- adopt the replacement of drupal_rebuild() into Drush - which is babysitting and I don't imagine you'll be happy with that, @weitzman 
- remove the `final` from CacheCommands -- I don't understand why it's been added
- make various methods in CacheCommands static so they can be called from outside it -- but it's going to be quite a few methods, basically everything that is called in https://github.com/joachim-n/drupal-core-development-project/blob/main/drush/Commands/core_development/DevelopmentProjectCommands.php

**Comment by weitzman** [comment]
I don't see any code in Drush that needs to be made static in order for your new cache:rebuild to work.

**Comment by joachim-n** [comment]
There are calls to:

$this->getAppRoot
$this->loadDrupalAutoloader

If I change DevelopmentProjectCommands so it no longer inherits from CacheCommands, how do I make those calls?

($this->logger I can get with inflection, IIRC)

**Comment by weitzman** [comment]
1. getAppRoot() is your code, right? No Drush change needed.
2. loadDrupalAutoloader() no longer exists. you can get the autoloader during the new create() method as CacheCommands does -  https://github.com/drush-ops/drush/blob/12.x/src/Commands/core/CacheCommands.php#L65. Docs about create()  are at https://www.drush.org/12.x/dependency-injection/#create-method
3. Every command extends the DrushCommands class. Thats gives you the logger() method that gets you a logger object.

**Comment by weitzman** [comment]
if you are trying to support Drupal 9 in same branch, you will want to change your composer constraint to be drush/drush:^11.6 || 12. 11.6 is where we backported create() approach. If just Drupal 10+, I'd suggest requiring Drush 12.

**Comment by joachim-n** [comment]
Hurrah, I've got it working! New release of joachim-n/drupal-core-development-project coming.

Thanks for your help @weitzman & @greg-1-anderson!

**Comment by weitzman** [comment]
FYI We moved cache:rebuild to a new CacheRebuildCommands file.
